stan/math/rev/mat/fun/multiply.hpp

```diff
@@ -59,80 +451,75 @@ namespace stan {
     }
 
     /**
-     * Return the product of the specified matrices.  The number of
-     * columns in the first matrix must be the same as the number of rows
-     * in the second matrix.
-     * @param[in] m1 First matrix.
-     * @param[in] m2 Second matrix.
-     * @return The product of the first and second matrices.
-     * @throw std::domain_error if the number of columns of m1 does not match
-     *   the number of rows of m2.
+     * Return the product of two matrices.
+     * @tparam TA scalar type matrix A 
+     * @tparam RA Rows matrix A
+     * @tparam CA Columns matrix A
+     * @tparam TB scalar type matrix B 
+     * @tparam RB Rows matrix B
+     * @tparam CB Columns matrix B
+     * @param[in] A Matrix.
+     * @param[in] B Matrix.
+     * @return Product of scalar and matrix.
      */
-    template<typename T1, int R1, int C1, typename T2, int R2, int C2>
+    template <typename TA, int RA, int CA, typename TB, int CB>
     inline typename
-    boost::enable_if_c< boost::is_same<T1, var>::value ||
-    boost::is_same<T2, var>::value,
-                        Eigen::Matrix<var, R1, C2> >::type
-      multiply(const Eigen::Matrix<T1, R1, C1>& m1,
-               const Eigen::Matrix<T2, R2, C2>& m2) {
-      check_multiplicable("multiply",
-                          "m1", m1,
-                          "m2", m2);
-      Eigen::Matrix<var, R1, C2> result(m1.rows(), m2.cols());
-      for (int i = 0; i < m1.rows(); i++) {
-        typename Eigen::Matrix<T1, R1, C1>::ConstRowXpr crow(m1.row(i));
-        for (int j = 0; j < m2.cols(); j++) {
-          typename Eigen::Matrix<T2, R2, C2>::ConstColXpr ccol(m2.col(j));
-          if (j == 0) {
-            if (i == 0) {
-              result(i, j) = var(new dot_product_vari<T1, T2>(crow, ccol));
-            } else {
-              dot_product_vari<T1, T2> *v2
-                = static_cast<dot_product_vari<T1, T2>*>(result(0, j).vi_);
-              result(i, j)
-                = var(new dot_product_vari<T1, T2>(crow, ccol, NULL, v2));
-            }
-          } else {
-            if (i == 0) {
-              dot_product_vari<T1, T2> *v1
-                = static_cast<dot_product_vari<T1, T2>*>(result(i, 0).vi_);
-              result(i, j)
-                = var(new dot_product_vari<T1, T2>(crow, ccol, v1, NULL));
-            } else /* if (i != 0 && j != 0) */ {
-              dot_product_vari<T1, T2> *v1
-                = static_cast<dot_product_vari<T1, T2>*>(result(i, 0).vi_);
-              dot_product_vari<T1, T2> *v2
-                = static_cast<dot_product_vari<T1, T2>*>(result(0, j).vi_);
-              result(i, j)
-                = var(new dot_product_vari<T1, T2>(crow, ccol, v1, v2));
-            }
-          }
+    boost::enable_if_c<boost::is_same<TA, var>::value ||
+                        boost::is_same<TB, var>::value,
```

Comment:
This isn't R, so we can start lines with operators (like `||`) in the usual way that math is written.  The reason is that it's easier to scan what's going on because we have the operators starting lines, and start of line is easier to see than end of line.